zkvm/src/contract.rs

```diff
@@ -0,0 +1,185 @@
+use crate::encoding;
+use crate::encoding::Subslice;
+use crate::errors::VMError;
+use crate::predicate::Predicate;
+use crate::txlog::{TxID, UTXO};
+use crate::types::{Commitment, Data, Value};
+use crate::vm::VariableCommitment;
+
+/// Prefix for the data type in the Output Structure
+pub const DATA_TYPE: u8 = 0x00;
+
+/// Prefix for the value type in the Output Structure
+pub const VALUE_TYPE: u8 = 0x01;
+
+/// A ZkVM contract that holds a _payload_ (a list of portable items) protected by a _predicate_.
+#[derive(Debug)]
+pub struct Contract {
+    pub(crate) payload: Vec<PortableItem>,
+    pub(crate) predicate: Predicate,
+}
+
+/// Representation of items that can be stored within outputs and contracts.
+#[derive(Debug)]
+pub enum PortableItem {
+    Data(Data),
+    Value(Value),
+}
+
+#[derive(Clone, Debug)]
+pub struct Input {
+    pub contract: FrozenContract,
+    pub utxo: UTXO,
+    pub txid: TxID,
+}
+
+/// Representation of a Contract inside an Input that can be cloned.
+#[derive(Clone, Debug)]
+pub struct FrozenContract {
+    pub(crate) payload: Vec<FrozenItem>,
+    pub(crate) predicate: Predicate,
+}
+
+/// Representation of a PortableItem inside an Input that can be cloned.
+#[derive(Clone, Debug)]
+pub enum FrozenItem {
+    Data(Data),
+    Value(FrozenValue),
+}
+
+/// Representation of a Value inside an Input that can be cloned.
+/// Note: values do not necessarily have open commitments. Some can be reblinded,
+/// others can be passed-through to an output without going through `cloak` and the constraint system.
+#[derive(Clone, Debug)]
+pub struct FrozenValue {
+    pub(crate) qty: Commitment,
+    pub(crate) flv: Commitment,
+}
+
+impl Contract {
+    pub fn min_serialized_length(&self) -> usize {
+        let mut size = 32 + 4;
+        for item in self.payload.iter() {
+            match item {
+                PortableItem::Data(d) => size += 1 + 4 + d.min_serialized_length(),
+                PortableItem::Value(_) => size += 1 + 64,
+            }
+        }
+        size
+    }
+
+    /// Half-way to encoding the contract
+    pub fn to_frozen(self, commitments: &Vec<VariableCommitment>) -> FrozenContract {
```

Comment:
Maybe it will be clearer if the whole freezing/unfreezing logic is kept in two thick methods in VM (for freezing and unfreezing), without smaller methods that receive pieces of vm state from outside.